modules/table/src/main/java/org/apache/ignite/internal/table/distributed/storage/InternalTableImpl.java

```diff
@@ -486,13 +476,54 @@ private <R> CompletableFuture<R> enlistWithRetry(
         return result;
     }
 
+    private <R> CompletableFuture<R> trackingInvoke(
+            InternalTransaction tx,
+            int partId,
+            Function<Long, ReplicaRequest> mapFunc,
+            boolean full,
+            IgniteBiTuple<ClusterNode, Long> primaryReplicaAndTerm,
+            @Nullable BiPredicate<R, ReplicaRequest> noOpChecker
+    ) {
+        ReplicaRequest request = mapFunc.apply(primaryReplicaAndTerm.get2());
+
+        boolean write = request instanceof SingleRowReplicaRequest && ((SingleRowReplicaRequest) request).requestType() != RW_GET
+                || request instanceof MultipleRowReplicaRequest && ((MultipleRowReplicaRequest) request).requestType() != RW_GET_ALL;
+
+        if (write && !full) {
+            // Track only write requests from explicit transactions.
```

Comment:
Why we track only write requests. AFAIK we should fail reading operations within finished transactions as well as modifications ones. WDYT?

Reply:
Tracking is required only for wait replication on commit purposes.